### PR TITLE
chore: optimize long list

### PR DESCRIPTION
**Pull request by SoloJiang**
*Before* submitting a pull request, please make sure the following is done...

1. Fork the repo and create your branch from `master`.
2. If you've added code that should be tested, add tests!
3. If you've changed APIs, update the documentation.
4. Ensure the test suite passes (`npm test`).
5. Make sure your code lints (`npm run lint`) - we've done our best to make sure these rules match our internal linting guidelines.

### CODE REVIEW

File: packages/jsx2mp-runtime/src/component.js [hunk 1]
```diff
@@ -269,3 +283,10 @@ function diffProps(prev, next) {
   }
   return false;
 }
+
+function diffArray(prev, next) {
+  if (!Array.isArray(prev)) return false;
+  // Only concern about list append case
+  if (prev.length === next.length) return false;
+  return next.slice(0, prev.length).every((val, index) => prev[index] === val);
```
Comment (wssgcg1213) [inline packages/jsx2mp-runtime/src/component.js]: `next.slice(0)` 就可以

File: packages/jsx2mp-runtime/src/component.js [hunk 2]
```diff
@@ -81,22 +81,36 @@ export default class Component {
...
      * */
-    this._internal.setData(data, () => {
-      this.__updating = false;
-    });
+    if (this._internal.$spliceData) {
+      const useSpliceData = {};
+      const useSetData = {};
+      for (let key in data) {
+        if (Array.isArray(data[key]) && diffArray(this.state[key], data[key])) {
+          useSpliceData[key] = [this.state[key].length, 0, ...data[key].slice(this.state[key].length)];
```
Comment (wssgcg1213) [inline packages/jsx2mp-runtime/src/component.js]: concat 会不会快一点?

File: packages/jsx2mp-runtime/src/component.js (hunk 1 above)
Comment (SoloJiang) [reply packages/jsx2mp-runtime/src/component.js]: 这里第二个参数是 prev.length 不是 next.length

Comment (wssgcg1213) [reply packages/jsx2mp-runtime/src/component.js]: 恩 看到了

File: packages/jsx2mp-runtime/src/component.js [hunk 3]
```diff
@@ -81,22 +81,42 @@ export default class Component {
...
+      const useSetData = {};
+      for (let key in data) {
+        if (Array.isArray(data[key]) && diffArray(this.state[key], data[key])) {
+          useSpliceData[key] = [this.state[key].length, 0].concat(data[key].slice(this.state[key].length));
+        } else {
+          if (diffData(this.state[key], data[key])) {
+            useSetData[key] = data[key];
+          }
+        }
+      }
```
Comment (wssgcg1213) [inline packages/jsx2mp-runtime/src/component.js]: 这块可以考虑抽一个方法

File: packages/jsx2mp-runtime/src/component.js [hunk 4]
```diff
@@ -269,3 +289,30 @@ function diffProps(prev, next) {
...
+    if (prevKeys.length !== nextKeys.length) return true;
+    if (prevKeys.length === 0) return false;
+    return !prevKeys.some(key => prevData[key] === nextData[key] );
+  } else {
+    return prevData !== nextData;
+  }
+}
+
+function isEmptyObj(obj) {
+  return Object.keys(obj).length === 0;
```
Comment (wssgcg1213) [inline packages/jsx2mp-runtime/src/component.js]: 这个比较慢吧  可以for in 一下, for in 里面直接返回false, 外层返回true

**Pull request merged**